arch/cpu/cc2538/lpm.h

```diff
@@ -139,6 +139,11 @@ void lpm_init(void);
  */
 void lpm_enter(void);
 
+/**
+ * Ensure that an interrupt will wake up the device again.
+ */
+void lpm_enter_pm2(void);
```

Comment:
This is something I have very strong reservations about.

You will see that the lpm module offers an `lpm_enter()` function, which will try to enter PM2 if i) it is enabled by the configuration and ii) if the system is overall in a good state to enter this PM. `lpm_enter()` makes sure to drop to a low-power mode gracefully, respecting the state of various on-chip peripherals as well as the state of the OS' scheduler at the point of dropping to low-power. As a very minor point, it also correctly handles the state of the ENERGEST module.

What this pull proposes is a new function that can be used to brute-forcefully enter PM2, merely for the sakes of resetting the state of the non-retention area of the CC2538's SRAM so a seed can be generated. There does not seem to be any consideration as to what other tasks may have been scheduled to happen, how long after the call to generate the seed. Call this function while the native USB driver is in use and you will immediately cause a USB device disconnect from the host (`lpm_enter()` will pick up the fact that the USB is active and it will not disrupt it).

What you want to be doing here is collect randomness from the state of the SRAM non-retention area opportunistically. The design I would recommend would be to add hooks to `lpm_enter()` and `lpm_exit()`. Upon entering a low power mode, check if it's gonna be PM2. If so, then upon wakeup inspect the state of the SRAM and gather randomness. Latch it and use it when required. Fail gracefully if sufficient randomness has not been generated.

Have a look at the newer and more advanced LPM implementation of the CC13x0/CC26x0. See how the LPM module notifies all interested parties before transition to a low-power state / upon wakeup. A similar approach would be recommended here.

Reply:
I am now reusing `lpm_enter`, but could not avoid some suboptimal adaptations of `lpm.c`.

Reply:
UPDATE: I am now getting away without modifications to `lpm.c` ...